### PR TITLE
THREESCALE-7405: Update Rails to 6.0

### PR DESCRIPTION
**Pull request by jlledom**
**What this PR does / why we need it**:

We need to upgrade to Rails 6 because our current Ruby and Rails versions have reached end of life.

This supersedes https://github.com/3scale/porta/pull/3339. That PR is a mess, some commits are repeated and some others are not relevant anymore, as their code is already backported to master. The changes here are exactly the same but the git history is cleaner.

**Which issue(s) this PR fixes** 

[THREESCALE-7405](https://issues.redhat.com/browse/THREESCALE-7405)

**Verification steps** 

Everything should work. Tests should pass.

**Special notes for your reviewer**:

In order to make it easier to review, this PR consists on 7 commits that address a different issue each one. Instead of reviewing the `Files changed` tab, where everything is mixed and out of context, I recommend reviewing each one of these commits individually, no matter the order. If there's a particular change which is hard to understand, https://github.com/3scale/porta/pull/3339 and its branch can be used as reference to track the changes and check the useful info on some commit messages there.

### CODE REVIEW

File: lib/developer_portal/lib/cms/toolbar.rb
Comment (mayorova) [inline lib/developer_portal/lib/cms/toolbar.rb:71]: Typo?

File: openshift/system/config/cache_store.yml
Comment (mayorova) [inline openshift/system/config/cache_store.yml:4]: Is this required? Can we set it as some default value? Just wondering, because if this is required, we need to update SaaS configs also

File: app/lib/liquid/template_support.rb
Comment (mayorova) [inline app/lib/liquid/template_support.rb:58]: I don't really understand this change :upside_down_face: but it seems that it all works :grimacing:

File: lib/developer_portal/lib/cms/toolbar.rb
Comment (jlledom) [reply lib/developer_portal/lib/cms/toolbar.rb:71]: Yes!

File: lib/developer_portal/lib/liquid/template/view.rb
Comment (nidhi-soni1104) [inline lib/developer_portal/lib/liquid/template/view.rb:6]: what is this change for?

File: openshift/system/config/cache_store.yml
Comment (jlledom) [reply openshift/system/config/cache_store.yml:4]: I think so. The digest algorithm was being configured through a monkey patch I removed, so it needs to be configured somehow. Maybe there are other options like setting it from an initializer or applicaton.rb, but I think config files are the best option.

File: app/lib/liquid/template_support.rb
Comment (jlledom) [reply app/lib/liquid/template_support.rb:58]: I've been checking this. It seems we need a monkey patch to render liquid templates in the dev. portal. The upgrade broke the implementation and the patch needed to be adapted. The new implementation is easy, when it's rendering a liquid template, use the liquid renderer, otherwise use the default Rails renderer.

I'd say there must be a way to fix it without the monkey patch, but that's out of scope here, I think.

See https://github.com/3scale/porta/commit/775bbc4fb4da3f7d2a410cacb17bcdc0d1af23fd and https://github.com/3scale/porta/commit/540100a053516afc44d87123c0137f9542056417

File: lib/developer_portal/lib/liquid/template/view.rb
Comment (jlledom) [reply lib/developer_portal/lib/liquid/template/view.rb:6]: See the log here: https://github.com/3scale/porta/commit/a6d7ae2238c256e7b79352f44a5adf334dcfc487

It's required because the initializer for the parent class changed, for instance now a `locals` param is mandatory, so we have to adapt to it.

File: openshift/system/config/cache_store.yml
Comment (mayorova) [reply openshift/system/config/cache_store.yml:4]: What monkey patch was this? Also, I can't find any reference to this setting in Rails documentation... 
But in any case, if this is the default option that we are not planning to change across environments, I'd really prefer to have is somewhere here: https://github.com/3scale/porta/blob/dbdf5b570a960982e841b9c0b03f60e29e613bb6/config/application.rb#L189

File: app/lib/liquid/template_support.rb
Comment (mayorova) [reply app/lib/liquid/template_support.rb:58]: Thanks for the clarification!

File: openshift/system/config/cache_store.yml
Comment (akostadinov) [reply openshift/system/config/cache_store.yml:4]: I think @jlledom doesn't mean monkey patch but this:

https://github.com/3scale/porta/pull/2971/files#diff-c1fd91cb1911a0512578b99f657554526f3e1421decdb9e908712beab57e10f9R51-R52

Comment (mayorova) [reply openshift/system/config/cache_store.yml:4]: If that's the case - we still have it here: https://github.com/3scale/porta/pull/3456/files#diff-c1fd91cb1911a0512578b99f657554526f3e1421decdb9e908712beab57e10f9R71-R73

I guess that applies for all ActiveSupport, cache included? (although, no idea)

Comment (jlledom) [reply openshift/system/config/cache_store.yml:4]: Sorry, I meant this: https://github.com/3scale/porta/commit/442b606278e8c632dfdee12352e298e4ebb89361#diff-ab512545d0eb817a27d443feddbdcf22962c873d891a20c8cc7fe63e5b0e9656L13

The options given to the config file are read from: 

https://github.com/3scale/porta/blob/dbdf5b570a960982e841b9c0b03f60e29e613bb6/config/application.rb#L189

so we end up calling this:

`config.cache_store = [:mem_cache_store, ["localhost:11211"], {:digest_class=>Digest::SHA256}]`

@mayorova If you really don't like this, I would prefer to set the digest in an initializer (I guess it's possible). But I would avoid adding more complexity to the cache section in `application.rb`

@akostadinov regarding this: https://github.com/3scale/porta/pull/2971/files#diff-c1fd91cb1911a0512578b99f657554526f3e1421decdb9e908712beab57e10f9R51-R52

`ActiveSupport::Digest.hash_digest_class = OpenSSL::Digest::SHA256`

That's not related. Setting `hash_digest_class` doesn't update the cache client digest, which is `MD5` by default.

```
[4] pry(main)> Rails.cache
=> #<ActiveSupport::Cache::MemCacheStore:0x000000000bd3e428
 @data=
  #<Dalli::Client:0x000000000bd3e1d0
   @key_manager=
    #<Dalli::KeyManager:0x000000000bd3e0b8
     @key_options={:digest_class=>Digest::MD5},
     @namespace=nil>,
   @normalized_servers=["localhost:11211"],
   @options={},
   @ring=nil>,
```

Comment (jlledom) [reply openshift/system/config/cache_store.yml:4]: @mayorova these are the docs for the Dalli Client, where `:digest_class` is defined:

https://github.com/petergoldstein/dalli/wiki/Dalli::Client-Options

Comment (mayorova) [reply openshift/system/config/cache_store.yml:4]: Thank you for the info @jlledom !

Well if both you and @akostadinov prefer to have it in the config file, that's fine. 

I just think if that's a "fixed" value that is not going to change, having it as a default value (maybe keep the option for overriding it via the config file) is more practical, so we don't need to carry it over to all the configuration files.

Comment (jlledom) [reply openshift/system/config/cache_store.yml:4]: I've been trying. The digest can't be overridden from an initializer, but it can be easily overridden in `application.rb`. I don't like it, but it's only one line. Anyway I think we should keep it in the configuration files, because it's configuration. Customers can be using other cache stores for which `:digest_class` means nothing (It's a Dalli exclusive option). I don't think it's a fixed value, it's our config.

Comment (mayorova) [reply openshift/system/config/cache_store.yml:4]: @akostadinov your opinion on :point_up: ?

I understand that @jlledom suggests just to set `:digest_class: !ruby/class Digest::SHA256` in the `cache_store.yml` config file.

I suggest to make this setting optional in the config file, and set the default value in `config/application.rb` (so, if you don't set anything in the config file `Digest::SHA256` will be used, but you can override it)

Comment (akostadinov) [reply openshift/system/config/cache_store.yml:4]: IIRC MD5 was hardcoded in dalli. And we better update it now we don't support RUby 2.4 anymore. @jlledom are you sure your setting makes any difference for our version of dalli?

Comment (akostadinov) [reply openshift/system/config/cache_store.yml:4]: I have no hard opinion where this configuration should be set. We only really care in FIPS mode where MD5 will fail anyway. This configuration should not be critical elsewhere.

Comment (jlledom) [reply openshift/system/config/cache_store.yml]: In order to unlock this, I pushed https://github.com/3scale/porta/pull/3456/commits/982b1cae28b1875facee10df996951a0801ed295. However, I found that the Rails cache management in this branch seems to be broken, not all cache stores work. I'll check this.

File: features/support/capybara.rb
Comment (jlledom) [reply features/support/capybara.rb:31]: yep, it fixed the oracle tests

File: app/models/settings.rb
Comment (akostadinov) [inline app/models/settings.rb]: Do we really need this? Is it too bad to call the proper method?

File: config/initializers/new_framework_defaults_6_0.rb
Comment (akostadinov) [inline config/initializers/new_framework_defaults_6_0.rb]: Will we need special precautions to handle migration? e.g. first drain the deliveries and only then deploy new version?

File: Gemfile.lock
Comment (akostadinov) [inline Gemfile.lock:934]: this seems like  big leap, is it necessary to do together with all other rails things?

File: Gemfile.lock
Comment (akostadinov) [inline Gemfile.lock:1077]: We may have issues with prod builds if we change bundler version. If we have to change, better update to 2.4 latest so that we have a significant improvement, instead of doing a very minor update.

File: config/application.rb
Comment (akostadinov) [inline config/application.rb:69]: I assume these two keep old defaults, correct?

Comment (akostadinov) [reply config/application.rb:69]: hmm, I see that these were already true in 5.2 https://guides.rubyonrails.org/configuring.html#default-values-for-target-version-5-2

Or do I miss where we changed them before?

They seem to be safe to update to `true` though. I guess only cookies will be invalidated, but in the past we considered that alright (for the FIPS update we had a similar situation).

File: config/environments/development.rb
Comment (akostadinov) [inline config/environments/development.rb:23]: The default stopped working after the update? Either case, I'd suggest that we allow anything `.*\.localhost`

File: config/initializers/mime_types.rb
Comment (akostadinov) [inline config/initializers/mime_types.rb:10]: I hate to nitpick for `isse`

File: app/models/email_configuration.rb
Comment (akostadinov) [inline app/models/email_configuration.rb:21]: need to change this? I remember some weird issues reported for some rails versions of difference between `before_save` and `before_create` so prefer to keep as is if we have no important reason

File: app/models/cinstance.rb
Comment (akostadinov) [inline app/models/cinstance.rb:39]: same here

File: app/models/settings.rb
Comment (jlledom) [reply app/models/settings.rb]: No, that's removed in the later cleanup commit: https://github.com/3scale/porta/pull/3456/commits/eb2346546cd7560f201510c26e8ec38b5fd52b6d#diff-6f750351adedd314c4f977fe861185a6f544b045e3d4969724786216a7dce38b

File: app/lib/finance/api/response.rb
Comment (akostadinov) [inline app/lib/finance/api/response.rb]: remember why do we need this?

File: app/workers/pdf_report_worker.rb
Comment (akostadinov) [inline app/workers/pdf_report_worker.rb:46]: can we remove this require? Seems redundant given `Report` should be auto-loaded

File: Gemfile.lock
Comment (akostadinov) [reply Gemfile.lock:934]: I see this was required because of `Add support for Rails 6.0.`

Going through the migration guide
https://github.com/CanCanCommunity/cancancan/blob/develop/docs/migrating.md#from-2x-to-3x

File: config/initializers/new_framework_defaults_6_0.rb
Comment (jlledom) [reply config/initializers/new_framework_defaults_6_0.rb]: I've been checking this. We don't need to take any action because old Rails 5.2 mailer jobs can be performed by Rails 6. This is what I found:

- They added a breaking change to `DeliveryJob` (https://github.com/rails/rails/pull/34367/commits/60339da5bcb76489576321fa12e665f176d8d692), then reverted it (https://github.com/rails/rails/pull/34591/commits/f5050d998def98563f8fa4b381c09f563681f159). So Rails 5 mailer jobs can be performed by Rails 6, but with a deprecation warnnig, see (https://github.com/rails/rails/pull/34591)
- All actions that enqueued a `DeliveryJob` or `Parameterized::DeliveryJob` ini Rails 5, now enqueue a `MailDeliveryJob` in Rails 6.
  - For instance, sending a message as provider to all users.
- This is what I tried locally:
  1. Send an email as provider to all users from Rails 5. Sidekiq is not running.
     1. Rails 5 enqueues a `MessageWorker` job.
     2. Switch to Rails 6.
     3. Run Sidekiq -> The job completes successfully.
  2. From the Rails 5 console:
     1. Sidekiq is not running. Run:
     ```
     [1] pry(main)> mes = Message.last
     [2] pry(main)> PostOffice.message_notification(mes, mes.recipients.first).deliver_later
     ```
     2. Rails 5 enqueues an `ActionMailer::DeliveryJob`.
     3. Switch to Rails 6.
     4. Run Sidekiq -> The job completes successfully, but with a deprecation warning.

File: Gemfile.lock
Comment (jlledom) [reply Gemfile.lock:1077]: I reverted this

File: config/application.rb
Comment (jlledom) [reply config/application.rb:69]: Those two where introduced in `5.2`, but we were using `5.1` defaults, so they were not set in porta. Now we use `6.0` defaults, those two are set to `true`, so I need to manually set them back to `false` to keep old behavior. Another question is if we want to keep the old behavior. I think we probably don't. WDTY?

Comment (mayorova) [reply config/application.rb:69]: I think we can use the newer defaults for `use_authenticated_cookie_encryption` and `use_authenticated_message_encryption`, and keep the overridden behavior for `to_time_preserves_timezone`

Comment (akostadinov) [reply config/application.rb:69]: agreed new defaults should be fine, whether to do now or after, is another question I have no hard opinion about

Comment (jlledom) [reply config/application.rb]: Ok, I removed those two lines.

File: config/initializers/mime_types.rb
Comment (jlledom) [reply config/initializers/mime_types.rb:10]: `unfortunatelly` there was another typo in that line. Fixed.

File: config/environments/development.rb
Comment (jlledom) [reply config/environments/development.rb:23]: It sees we need it. I tried removing it and porta doesn't load. I changed the regexp to `/.+\.localhost/`

File: app/workers/pdf_report_worker.rb
Comment (jlledom) [reply app/workers/pdf_report_worker.rb]: Removed

File: app/models/email_configuration.rb
Comment (mayorova) [reply app/models/email_configuration.rb:21]: You are right @akostadinov 

I've looked into it, and the "weird issues" is basically that `before _save` just ignores the `on:` option: https://github.com/rails/rails/issues/17622

This was "fixed" in Rails 6.0 by disallowing the `on:` options: https://github.com/rails/rails/pull/30919

Now it throws
```
Uncaught exception: Unknown key: :on. Valid keys are: :if, :unless, :prepend
```

So yeah, this change is required, the question is - should we just keep `before_save` (because I guess that would be keeping the existing behavior), or switch to `before_create` (which I assume was the intended - although not working - behavior)?

Comment (mayorova) [reply app/models/email_configuration.rb:21]: In this particular case I think we can use just `before_save` - in case we have some older models that did not have `tenant_id` initially, they will still get it on update.

File: app/models/cinstance.rb
Comment (mayorova) [reply app/models/cinstance.rb:39]: I think here it makes sense to change to `after_create`, although not sure if that would be breaking something in the existing behavior :thinking:

File: app/models/email_configuration.rb
Comment (jlledom) [reply app/models/email_configuration.rb:21]: we should use `before_create` because `before_save` would call `set_tenant_id` also when `#update`. Is `before_create` not working? Why?

File: app/models/cinstance.rb
Comment (jlledom) [reply app/models/cinstance.rb:39]: If I understood this correctly, `after_save` ignores the `on:` option in rails 5. So right now in master `create_first_key` is being called every time the application is `#updated`. `after_create` seems the correct behavior for me.

File: app/models/email_configuration.rb
Comment (mayorova) [reply app/models/email_configuration.rb:21]: @jlledom what I meant is that right now I think `before_save :set_tenant_id` is the equivalent of `before_save :set_tenant_id, on: :create`

So, by keeping `before_save` we're just maintaining the existing logic.

And my point is that I think it's OK if we call this method on `#update` also, as it will do nothing if `tenant_id` already is set.

File: app/lib/finance/api/response.rb
Comment (jlledom) [reply app/lib/finance/api/response.rb]: Thales did it [here](https://github.com/3scale/porta/pull/3339/commits/f15bfd6b604b480575456cadb0ae114d81a1bfd5#r1238505018).

I think this was a leftover or unused file, and I removed it in the later cleanup commit, [here](https://github.com/3scale/porta/pull/3456/commits/eb2346546cd7560f201510c26e8ec38b5fd52b6d#diff-2420eda950edd6e9d6e91336681bff689d2cd4a8d9d81118ab8ea3ac08ae2369)

File: app/models/cinstance.rb
Comment (jlledom) [reply app/models/cinstance.rb:39]: I'm leaving it as `after_create`, I think it's better.

File: app/models/email_configuration.rb
Comment (jlledom) [reply app/models/email_configuration.rb]: I changed this to `before_save`

File: Gemfile.lock
Comment (akostadinov) [reply Gemfile.lock:934]: See #3481

File: app/models/account.rb
Comment (akostadinov) [inline app/models/account.rb:200]: I assume such changes are because of https://github.com/rails/rails/pull/35198

File: Gemfile.lock
Comment (jlledom) [reply Gemfile.lock:934]: I merged master which already contains this change. No N+1 errors were detected in tests.

**Review by akostadinov — APPROVED** [review]
Thank you for the really great work!

Approving with two notes:
* need to merge after we validate some recent changes in master
* will need regenerating the schema files right before merging

File: Gemfile
Comment (akostadinov) [inline Gemfile:17]: why? how would this be useful in production?

Comment (jlledom) [reply Gemfile:17]: I moved it out of the development group because it's being required [here](https://github.com/3scale/porta/pull/3456/files#diff-441bbd2acf3b2eab84a98d917f574109eafd6c54cbce74a2379487f9f67920a2R4) and that's the reason the community image fails to build.

But you're right, the best solution is to not call `require 'bootsnap/setup'` when rails environment is `:production`.

Comment (mayorova) [reply Gemfile]: What's the problem though having it in production?
I understand that it's Rails who adds this by default, so I guess there might be some benefit for production as well?

Comment (jlledom) [reply Gemfile]: I think it's better to not have it in production because it uses a `/tmp` folder in disk as cache, and that cache is never erased. So we need to maintain that folder, either keeping it in the container filesystem or mounting a volume for it, in case the container is read-only. And it's only advantage is increasing boot speed by a few seconds, which we won't really notice since the whole rollout process in OCP takes like 5-7 minutes.

File: app/lib/plan_rule_loader.rb
Comment (mayorova) [inline app/lib/plan_rule_loader.rb:29]: Is this needed? I understand that now all configs pulled in by `config_for` are already deep-symbolized.

File: test/unit/apicast/provider_source_test.rb
Comment (mayorova) [inline test/unit/apicast/provider_source_test.rb:35]: Hmmm, why this?...

File: app/lib/plan_rule_loader.rb
Comment (jlledom) [reply app/lib/plan_rule_loader.rb:29]: I've been trying and it seems it's actually required. For instance, [this call](https://github.com/3scale/porta/blob/THREESCALE-7405-rails-6/config/application.rb#L243) to `config_for` returns string keys. And [the call](https://github.com/3scale/porta/blob/THREESCALE-7405-rails-6/config/application.rb#L228) that reads the plan rules uses `try_config_for` which calls `symbolize_keys`, but not `deep_*` so only the top level keys are symbolized.

File: test/unit/apicast/provider_source_test.rb
Comment (jlledom) [reply test/unit/apicast/provider_source_test.rb:35]: Here's the commit: https://github.com/3scale/porta/commit/0a8faa77c583bc0e68b58a20bc7e68b29139c656

I tried the test and it fails if I remove that conversion to `iso8601`. It seems Rails' `as_json` changed the way it returns dates and that made the tests fail.

**Comment by jlledom** [comment]
> I guess this line is some old stuff: https://github.com/3scale/porta/blob/THREESCALE-7405-rails-6/.codeclimate.yml#L38

Yes, it refers to a file that doesn't exist anymore. Deleted.

File: app/lib/plan_rule_loader.rb
Comment (mayorova) [reply app/lib/plan_rule_loader.rb:29]: Hmmm, I don't get it... 

Accessing config using non-symbol keys is deprecated, but the keys are not symbolized automatically?... :thinking: 

Let me double-check what's going on there

File: test/unit/apicast/provider_source_test.rb
Comment (mayorova) [reply test/unit/apicast/provider_source_test.rb:35]: Yeah, what I mean is - if the test is failing now, does it mean that the application behavior has changed?? Maybe it means we need to fix something else??

File: app/lib/plan_rule_loader.rb
Comment (mayorova) [reply app/lib/plan_rule_loader.rb:29]: https://github.com/3scale/porta/pull/3501

Up to you if you want to merge this or not :) 

I just felt curious about this change.

Also, if I am wrong, please let me know.

File: Gemfile
Comment (akostadinov) [reply Gemfile]: The way it is required is incorrect as far as I remember. I think it never properly worked. I had it on the back of my mind to some day look at it. If necessary, stop requiring it at all and at some point, we will see how to properly enable it.

File: test/unit/apicast/provider_source_test.rb
Comment (akostadinov) [reply test/unit/apicast/provider_source_test.rb:35]: good point

File: Gemfile
Comment (jlledom) [reply Gemfile]: I don't know. It seems to me the porta boots much faster locally in Rails 6 than Rails 5. So it should be working fine.

File: test/unit/apicast/provider_source_test.rb
Comment (jlledom) [reply test/unit/apicast/provider_source_test.rb:35]: Let me check

Comment (mayorova) [reply test/unit/apicast/provider_source_test.rb:35]: Hm.... So, I don't really understand this, on line `:33` there is `service.stubs(updated_at: Time.now)` and I think this is actually what causes the issue.

I think it's because `updated_at` is `ActiveSupport::TimeWithZone` and it handles the equality with ISO8601 correctly, while `Time` doesn't.

But actually, I've dug into this class, and I think it's a deprecated functionality, which probably needs to be removed, so I am not bothered by it anymore :) 
In fact, I think I'll create a JIRA to remove the related classes.

Comment (mayorova) [reply test/unit/apicast/provider_source_test.rb:35]: Sorry, I was wrong, the endpoint that uses this class is still in use... 
But well, I am not sure to be honest whether this specific field `updated_at` is actually useful.

File: config/boot.rb
Comment (mayorova) [inline config/boot.rb:5]:
```suggestion
require 'bootsnap/setup' unless ENV['RAILS_ENV'] == 'production' # Speed up boot time by caching expensive operations.
```

Sorry, couldn't resist!

Comment (akostadinov) [reply config/boot.rb:5]: I like double quotes :) Changing between double and single when interpolation is added or removed seems to me unnecessary

Comment (mayorova) [reply config/boot.rb:5]: Well, It's for consistency... We don't have much of it within the whole project (while I think we should), but at least I'd prefer not to mix both styles within the same file. This file already uses single quotes, so I suggested changing it here also for consistency (within the file, and also within the line!).

File: test/unit/apicast/provider_source_test.rb
Comment (akostadinov) [reply test/unit/apicast/provider_source_test.rb:35]: I think it is much better to compare a string to a string instead of relying on implicit coercion. And we make sure that the actual format of the string is preserved across Rails and other libraries upgrades.

**Review by akostadinov — DISMISSED** [review]
Great Job!

File: config/boot.rb
Comment (jlledom) [reply config/boot.rb:5]: This is unacceptable, but accepted.

**Review by mayorova — DISMISSED** [review]
Awesome job @jlledom ! :trophy: :1st_place_medal:

**Review by mayorova — APPROVED** [review]

**Pull request merged**
